reactor-netty-core/src/main/java/reactor/netty/transport/TransportConnector.java

```diff
@@ -219,7 +220,35 @@ static void doConnect(SocketAddress remoteAddress, @Nullable Supplier<? extends
 				return monoChannelPromise;
 			}
 
+			if (config instanceof ClientTransportConfig) {
+				final ClientTransportConfig<?> clientTransportConfig = (ClientTransportConfig<?>) config;
+				if (clientTransportConfig.doOnResolve != null) {
+					clientTransportConfig.doOnResolve.accept(Connection.from(channel));
+				}
+			}
+
 			Future<SocketAddress> resolveFuture = resolver.resolve(remoteAddress);
+
+			if (config instanceof ClientTransportConfig) {
+				final ClientTransportConfig<?> clientTransportConfig = (ClientTransportConfig<?>) config;
+
+				if (clientTransportConfig.doOnResolveError != null) {
```

Comment:
I'd combine these two blocks (test+addListener) into a single listener, but I guess the cost of adding an extra listener is not that high

Reply:
This was originally a single listener.  However, @violetagg had a good suggestion that we do not want to register the listener unnecessarily when the hooks are not there.  I can do a single listener and achieve both requests; however, that would cause the code to look a little uglier (making some checks multiple times).  Here is how it would look like:

```java
if (config instanceof ClientTransportConfig) {
	final ClientTransportConfig<?> clientTransportConfig = (ClientTransportConfig<?>) config;

	if (clientTransportConfig.doOnResolveError != null || clientTransportConfig.doAfterResolve != null) {
		resolveFuture.addListener((FutureListener<SocketAddress>) future -> {
			if (future.cause() != null) {
				if (clientTransportConfig.doOnResolveError != null) {
					clientTransportConfig.doOnResolveError.accept(Connection.from(channel), future.cause());
				}
			} else {
				if (clientTransportConfig.doAfterResolve != null) {
					clientTransportConfig.doAfterResolve.accept(Connection.from(channel), future.getNow());
				}
			}
		});
	}
}
```

I could potentially make `doOnResolveError` and `doAfterResolve` use `Optional` to eliminate some of the `if` checks, but that would cause a deviation from other hooks.  In addition, I am not sure if the code would be any nicer.

Reply:
yeah that's what I was getting at. seeing a snippet of code helps, indeed there's no huge gain in readability so let's keep as is 👍